### PR TITLE
Export ReduceMean/ReduceFrontMean/ReduceBackMean (Caffe2) to ReduceMean (ONNX).

### PR DESCRIPTION
**Pull request by xkszltl**
The second input (`lengths`) is not supported.

### CODE REVIEW

**Comment by xkszltl** [comment]
Test added.
CI failure looks unrelated.

**Comment by ezyang** [comment]
@yinghai could you please review this?

**Review by houseroad — COMMENTED** [review]
ReduceFrontMean and ReduceBackMean are deprecated now. Could you (also) add support for ReduceMean operator (https://github.com/pytorch/pytorch/blob/master/caffe2/operators/reduce_ops.cc#L221)?

**Comment by xkszltl** [comment]
@houseroad Added

File: caffe2/onnx/onnx_exporter.cc [hunk 1]
```diff
@@ -768,6 +772,79 @@ ConvertedResult OnnxExporter::CreateChannelShuffleNodes(
...
+        args.emplace(a.name(), &a);
+    }
+
+    std::vector<int64_t> axes;
+    int64_t keepdims = 1;
+
+    if (def.type() == "ReduceMean") {
+        // axes
+        {
+            const auto it = args.find("axes");
```
Comment (houseroad) [inline caffe2/onnx/onnx_exporter.cc]: using additional block to avoid dup names seems uncommon in PyTorch codebase. Could you remove this trick?

Comment (xkszltl) [reply caffe2/onnx/onnx_exporter.cc]: This provides more benefit than what you thought.
1. Use `it_suffix` is unnecessary and it makes code becomes longer.
2. With `const` we get SSA, which is easier to read for "local" var.
3. With `{}` I can get even more, i.e. terminate the life span earlier before reassignment. This narrows the "local" semantics and helps people navigating in the code.

Comment (houseroad) [reply caffe2/onnx/onnx_exporter.cc]: I didn't see the benefits here...

Especially, the logic here is so simple, and you can just reuse `it` name.

File: caffe2/onnx/onnx_exporter.cc [hunk 2]
```diff
@@ -768,6 +772,79 @@ ConvertedResult OnnxExporter::CreateChannelShuffleNodes(
...
+        // keepdims
+        {
+            const auto it = args.find("keepdims");
+            if (it != args.end()) {
+                keepdims = it->second->i();
+            }
+        }
+    } else {
+        // num_reduce_dim
+        {
```
Comment (houseroad) [inline caffe2/onnx/onnx_exporter.cc]: Also, the extra block is useless here. And it will make the coding style so inconsistent...

**Review by houseroad — COMMENTED** [review]
Please make sure the coding style is consistent with the whole file.

File: caffe2/onnx/onnx_exporter.cc (hunk 1 above)
Comment (xkszltl) [reply caffe2/onnx/onnx_exporter.cc]: Fine...Don't wanna drag this PR longer.
IMHO reuse is bad, and makes the first occurrence asymmetric to the rest.

**Comment by xkszltl** [comment]
Error unrelated.
CI tests failed to start.

```
DOCKER_IMAGE: 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda10-cudnn7-py3-gcc7:285
Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-cuda10-cudnn7-py3-gcc7:285 not found
Exited with code 1
```

**Review by houseroad — APPROVED** [review]
Looks good. Thanks.

**Pull request closed without merging**